Recon/src/Recon/OverlayProducer.cxx

```diff
@@ -124,199 +100,250 @@ void OverlayProducer::produce(framework::Event &event) {
+    }
+
     if (verbosity_ > 2) {
-      ldmx_log(debug) << "in loop: overlaying event " << iEv + 1 << " out of "
-                      << nEvsOverlay;
+      ldmx_log(debug) << "in loop: start of collection " << collName
+                      << "in loop: printing current sim event: ";
     }
+    ldmx_log(debug) << "in loop: size of sim hits vector " << collName << " is "
+                    << simHitsCalo.size();
+
+    // we don't need to touch the hard process sim hits, really... but we
+    // might need the simhits in the hit map.
+    if (needsContribsAdded || verbosity_ > 2) {
+      for (const ldmx::SimCalorimeterHit &simHit : simHitsCalo) {
+        if (verbosity_ > 2)
+          simHit.Print();
+
+        if (needsContribsAdded) {
+          // this copies the hit, its ID and its coordinates directly
+          hitMap[simHit.getID()] = simHit;
+        }
 
-    // an overlay event wide time offset to be applied to all its hits.
-    // TODO -- figure out if we should also randomly shift the time of the sim
-    // event (likely only needed if time bias gets picked up by BDT or ML by way
-    // of pulse behaviour)
-    float timeOffset = rndmTime_->Gaus(timeMean_, timeSigma_);
-    int bunchOffset = (int)rndmTime_->Uniform(
-	   	  -(nEarlier_+1) , nLater_+1);  // +1 to get inclusive interval
-    float bunchTimeOffset = bunchSpacing_ * bunchOffset;
-    timeOffset += bunchTimeOffset;
+      } // over calo simhit collection
+    }   // if needContribs or very verbose
+
+  } // over calo collections for sim event
+
+  /* ----------- now do the same with SimTrackerHits! ----------- */
+
+  // get the SimTrackerHit collections that we want to overlay, by looping
+  // over the list of collections passed to the producer : trackerCollections_
+  for (uint iColl = 0; iColl < trackerCollections_.size(); iColl++) {
+
+    std::string outCollName = trackerCollections_[iColl] + "Overlay";
+
+    simHitsTracker = event.getCollection<ldmx::SimTrackerHit>(
+        trackerCollections_[iColl], simPassName_);
+    trackerCollectionMap[outCollName] = simHitsTracker;
+
+    // the rest is printouts for debugging
+    ldmx_log(debug) << "in loop: size of sim hits vector "
+                    << trackerCollections_[iColl] << " is "
+                    << simHitsTracker.size();
 
     if (verbosity_ > 2) {
-      ldmx_log(debug) << "hit time offset in event " << iEv + 1 << " is  "
-                      << timeOffset << " ns";
-      ldmx_log(debug) << "bunch position offset in event " << iEv + 1 << " is  "
-                      << bunchOffset
-                      << ", leading to an additional time offset of "
-                      << bunchTimeOffset << " ns";
+      ldmx_log(debug) << "in loop: start of collection "
+                      << trackerCollections_[iColl];
+      ldmx_log(debug) << "in loop: printing current sim event: ";
+
+      for (const ldmx::SimTrackerHit &simHit : simHitsTracker) {
```

Comment:
If you are using this for loop just for printout, I would surround the whole thing with the `if (verbosity_ > 2)` condition. Otherwise, you are looping through the hits every event for no reason.

Reply:
Actually, the calorimeter and sim hit loops seem to be doing basically the same thing. I would pull this out into a function and just run all collections through it.  It will cut down on code repeat and improve readability.

Reply:
> Yeah, we've discussed the sim tracker vs sim calo hits treatment before, and initially I wanted a catch-all solution for this (more elegant, less error prone, all of that) but we came to the conclusion (in a sw dev meeting I think?) that it was not trivial to make a single function for it given the different types.

Yeah, this won't hold up the merge. Just wanted to note it.